[WIP] modified whitespacetokenizer to use torchtext vocab

parmeet opened the pull request:
This is WIP to use torchtext building blocks in WhitespaceTokenizer

Current Modifications:
- Replace python dictionary for member vocab with torchtext vocab object
- Modified API: Removed 'speed' keyword argument from __init__

parmeet committed:
modified whitespacetokenizer to use torchtext vocab

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,43 @@
 
         # load vocab
-        path_manager = PathManager()
+        ordered_dict = OrderedDict()
```
Can we use [build_vocab_from_text_file](https://github.com/pytorch/text/blob/master/torchtext/experimental/vocab.py#L22)?

parmeet replied:
Yes, now we are making use of it.

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,43 @@
+        token_ids = self.vocab(tokens)
+
+        if self.trainable:
+            unk_id = self.vocab.__getitem__('<unk>')
```
Can you use `unk_id = self.vocab(['<unk>'])` here. See the [doc](https://github.com/pytorch/text/blob/master/torchtext/experimental/vocab.py#L163).

parmeet replied:
> Please double check the new whitespacetokenizer is jitable in the context.
> 
> ```
> tokenizer = WhitespaceTokenizer(vocab_path)
> jit_tokenizer = torch.jit.script(tokenizer)
> ```

Yes, this is validated via doc_classification workflow (it seems to not working for non-jitable WhitespaceTokenizer)

zhangguanheng66 reviewed (commented):
Please double check the new whitespacetokenizer is jitable in the context.
```
tokenizer = WhitespaceTokenizer(vocab_path)
jit_tokenizer = torch.jit.script(tokenizer)
```

parmeet committed:
usage of sentenc_piece and build_vocab_from_text_file

parmeet commented:
Further modifications in commit2:

- Using build_vocab_from_text_file in WhitespaceTokenizer
- Now making use of basic_english_normalize inside WhitespaceTokenizer (needed by build_vocab_from_text_file)
- doc_classification workflow: DocClassificationDataModule setup is now using sentence_piece tokenizer and Vocab object from torchtex.experimental.transforms

zhangguanheng66 commented on stl_text/datamodule/doc_classification.py:
```diff
@@ -27,17 +51,23 @@ def __init__(self, data_path: str = 'glue_sst2_tiny', vocab_path: Optional[str]
+        else:
+            vocab = Vocab(VocabPybind([self.unk_token],self.unk_token))
+
+        vocab = vocab.to_ivalue()
```
I'd add [TODO] note here. We have a PR to remove the `to_ivalue` func.
https://github.com/pytorch/text/pull/1080

zhangguanheng66 commented on stl_text/datamodule/doc_classification.py:
```diff
@@ -27,17 +51,23 @@ def __init__(self, data_path: str = 'glue_sst2_tiny', vocab_path: Optional[str]
-        # use dummy text_transform and label_transform to get rid of pickling and torchscript issues
-        self.text_transform = WhitespaceTokenizer(vocab_path=self.vocab_path)
+        # usage of below tokenizer is same as above except that basic_english_normalize() is used as tokenizer
+        # self.text_transform = WhitespaceTokenizer(vocab_path=self.vocab_path)
```
We should find a way to support both tokenizers in the workflow. The ultimate solution will be `config`. For now, can we pass an argument to the func?

zhangguanheng66 commented on stl_text/datamodule/doc_classification.py:
```diff
@@ -9,6 +9,30 @@
+from torchtext.utils import download_from_url
+import torch.nn as nn
+
+class TextToIndices(nn.Module):
```
This is the same transform as [TextSequentialTransforms](https://github.com/pytorch/text/blob/master/torchtext/experimental/transforms.py#L170). Also you can use a lambda func.

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,46 @@
         super(WhitespaceTokenizer, self).__init__()
         self.trainable = trainable
-        self.speed = speed  # mock a real tokenizer: slowing down tokenization speed
+        self.tokenizer = transforms.basic_english_normalize().to_ivalue()
```
Do we need `basic_english_normalize` here? I think we can still use the `split` func as this is a white space tokenizer.

parmeet replied:
Agreed!

zhangguanheng66 reviewed (commented):
Please make sure that 1) the workflow works once you switch to the sentencepiece and vocab blocks in torchtext; 2) the components have JIT support.

parmeet committed:
- Added control flow for usage of tokenizer in doc_workflow.

- Restore split tokenizer in WhitespaceTokenizer

parmeet commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,52 @@
+class SplitTextModule(nn.Module):
+    def __init__(self):
+        super(SplitTextModule,self).__init__()
 
```
I am trying to see if explicitly creating Module for split() can be avoided. The reason I have to do this is build_vocab_from_text_file require jitable tokenizer.

zhangguanheng66 replied:
You can try [load_vocab_from_file](https://github.com/pytorch/text/blob/master/torchtext/experimental/vocab.py#L53) which simply loads the vocab column from a txt file.

parmeet committed:
minor syntactic change

parmeet committed:
using load_vocab_from_file in WhitespaceTokenizer

parmeet committed:
Modified pytext_vocab tokenizer to use torchtext vocab

added doc string for JIT code output

zhangguanheng66 commented on stl_text/datamodule/doc_classification.py:
```diff
@@ -27,17 +43,27 @@ def __init__(self, data_path: str = 'glue_sst2_tiny', vocab_path: Optional[str]
+                with path_manager.open(self.vocab_path, "r",encoding='utf-8') as f:
+                    vocab = build_vocab_from_text_file(f,torch.jit.script(tokenizer))
+            else:
+                vocab = Vocab(VocabPybind([self.unk_token],self.unk_token))
```
If `self.vocab_path` is not provided, we should check out the token->index dictionary in the sentencepiece model and use it in the vocab.

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,57 @@
 
 class WhitespaceTokenizer(nn.Module):
-    def __init__(self, vocab_path: Optional[str] = None, trainable=False, speed: int = 0):
+    """
```
you can follow the doc string format [here](https://github.com/pytorch/text/blob/master/torchtext/experimental/transforms.py#L77).

parmeet replied:
> you can follow the doc string format [here](https://github.com/pytorch/text/blob/master/torchtext/experimental/transforms.py#L77).

sure, thanks for the pointer. Do we still want to keep JIT output in the doc string?

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,57 @@
+            with path_manager.open(vocab_path, "r",encoding='utf-8') as f:
+                self.vocab = load_vocab_from_file(f,unk_token=self.unk_token)
+        else:
+            self.vocab = Vocab(VocabPybind([self.unk_token],self.unk_token))
```
Why do we need `Vocab(VocabPybind())`? Should we call `VocabPybind` to build a torchtext `Vocab`?

parmeet replied:
> Why do we need `Vocab(VocabPybind())`? Should we call `VocabPybind` to build a torchtext `Vocab`?

My bad! we could use vocab factory method here..

zhangguanheng66 commented on stl_text/ops/vocab/pytext_vocab.py:
```diff
@@ -16,8 +19,52 @@
+    jit_vocab = torch.jit.script(vocab)
+    print(jit_vocab.code)
+
+    >>>>>>>>>>>>>>>>>>>>>>>>>>OUTPUT of JIT code<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
```
same comment for the doc strings.

zhangguanheng66 commented on stl_text/ops/vocab/pytext_vocab.py:
```diff
@@ -34,42 +81,42 @@ def __init__(
+
     def forward(self, tokens: List[List[str]]) -> List[List[int]]:
-        tokens_idx = self.vocab.lookup_indices_2d(tokens)
+        tokens_idx = [self.vocab(token) for token in tokens]
```
If you are using torchtext Vocab, you can put a list of tokens directly to the vocab. See [here](https://github.com/pytorch/text/blob/master/torchtext/experimental/vocab.py#L163). Lookup line-by-line is generally faster than the lookup word-by-word.

parmeet replied:
> If you are using torchtext Vocab, you can put a list of tokens directly to the vocab. See [here](https://github.com/pytorch/text/blob/master/torchtext/experimental/vocab.py#L163). Lookup line-by-line is generally faster than the lookup word-by-word.

Right! The input is List of List of tokens. So we are doing line-by-line look up :)

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,57 @@
 
-    def forward(self, text: str) -> List[int]:
-        return self.numberize(self.tokenize(text))
+        self.vocab = self.vocab.to_ivalue() #Remove to_ivalue() PR: https://github.com/pytorch/text/pull/1080
```
On the eager mode, we suggest to use the pybind vocab (aka, not to call `to_ivalue`) for better performance. `to_ivalue` is called to switch to torchbind vocab before calling JIT.

zhangguanheng66 commented on stl_text/ops/vocab/pytext_vocab.py:
```diff
@@ -34,42 +81,42 @@ def __init__(
         # Make room for bos and eos from max_seq_len if true
         self.truncate_transform = TruncateTransform(max_seq_len - add_bos - add_eos)
 
+        self.vocab = self.vocab.to_ivalue() #Remove to_ivalue() PR: https://github.com/pytorch/text/pull/1080
```
On the eager mode, we suggest to use the pybind vocab (aka, not to call `to_ivalue`) for better performance. `to_ivalue` is called to switch to torchbind vocab before calling JIT.

parmeet replied:
> On the eager mode, we suggest to use the pybind vocab (aka, not to call `to_ivalue`) for better performance. `to_ivalue` is called to switch to torchbind vocab before calling JIT.

Right, i also notice similar behavior when doing the benchmarking...

zhangguanheng66 replied:
Same as my comment above.

parmeet committed:
Modified doc string for whitespacetokenizer and VocabTransform

Removed  VocabPyBind by vocab factory method

Added sentencepience_processor in doc_classification workflow

parmeet committed:
this file should have been part of previous commit

parmeet committed:
adding VocabTransform to package

zhangguanheng66 commented on stl_text/datamodule/doc_classification.py:
```diff
@@ -27,17 +49,27 @@ def __init__(self, data_path: str = 'glue_sst2_tiny', vocab_path: Optional[str]
-        # self.text_transform = sequential_transforms(tokenizer, vocab)
+        if self.tokenizer_type=='sentencepiece':
+            if self.vocab_path:
+                tokenizer = sentencepiece_tokenizer(download_from_url(PRETRAINED_SP_MODEL["text_unigram_25000"])).to_ivalue()
```
The workflow should still work even if you use the pybind version (a.k.a. without `to_ivalue`).

zhangguanheng66 commented on stl_text/datamodule/doc_classification.py:
```diff
@@ -27,17 +49,27 @@ def __init__(self, data_path: str = 'glue_sst2_tiny', vocab_path: Optional[str]
+                vocab = vocab.to_ivalue() #TODO Remove to_ivalue() PR: https://github.com/pytorch/text/pull/1080
+                self.text_transform = TextSequentialTransforms(OrderedDict([('tokenizer',tokenizer),('vocabulary',vocab)]))
+            else:
+                self.text_transform = sentencepiece_processor(download_from_url(PRETRAINED_SP_MODEL["text_unigram_25000"])).to_ivalue() 
```
Same here. The workflow should still work even if you use the pybind version (a.k.a. without to_ivalue).

zhangguanheng66 commented on stl_text/datamodule/doc_classification.py:
```diff
@@ -9,12 +9,34 @@
+    TextSequentialTransforms,
+)
+from torchtext.experimental.vocab import (
+    build_vocab_from_text_file,
```
Do we use `build_vocab_from_text_file` func in the file?

parmeet replied:
> Do we use `build_vocab_from_text_file` func in the file?

Not really now anymore. I guess we do not not need this since sentencepiece is already pre-trained. As you suggested previously, we could directly use sentencepiece_processor to get token ids directly from input text. I just l just cleaned up code a bit.

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,62 @@
+        else:
+            self.vocab = vocab(OrderedDict(),unk_token=self.unk_token)
+
+        self.vocab = self.vocab.to_ivalue() #Remove to_ivalue() PR: https://github.com/pytorch/text/pull/1080
```
Same as my comment above.

zhangguanheng66 commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,62 @@
-                    self.vocab[token] = len(self.vocab)
-                else:
-                    token_ids.append(self.vocab[self.unknown])
+        token_ids = self.vocab(tokens)
```
Not very sure. Seem that we should separate  the `training` process from the `numberize` func. Probably we can have another func to train the vocab. @hudeven any special usage cases to couple these two steps together?

hudeven replied:
Agree, let's have another func to train the vocab(e.g. build_vocab(data_file, tokenizer)). It's needed for translation task: https://github.com/hudeven/text/blob/master/examples/translation/train.py#L49

parmeet replied:
@hudeven @zhangguanheng66 I included the build_vocab function inside whitespacetokenizer so that train vocab can be separated from forward function. On a separate note, I realize that https://github.com/hudeven/text/blob/master/examples/translation/train.py#L88 is failing on my side.

parmeet committed:
some clean-up in doc_classification

parmeet committed:
adding functionality to build vocab from text file when using sentencepiece in doc_classification

Added vocab text files for testing purpose both for sentencepiece and whitespace tokenizers

parmeet committed:
seperating 'trainable' into seperate function in WhitespaceTokenizer

Accordingly modified translation task

parmeet committed:
removing debugger from main

cpuhrsch commented on stl_text/ops/tokenizers/whitespace_tokenizer.py:
```diff
@@ -1,44 +1,51 @@
+        return self.vocab(text.split())
+
+    @torch.jit.ignore
+    def build_vocab(self, data: List[str]) -> None:
```
What is the reasoning behind making this a method of the tokenizer?

parmeet replied:
> What is the reasoning behind making this a method of the tokenizer?

The original implementation of WhitespaceTokenizer included a trainable option that allow to build the vocab on the fly (https://github.com/hudeven/text/blob/9fc5b4962282d7e1dc8d5fa5eb3258f253981060/stl_text/ops/tokenizers/whitespace_tokenizer.py#L40). 

We discussed on possibility to separate "trainability" from querying (https://github.com/hudeven/text/pull/10#discussion_r563836230)

So as an initial attempt I just put this in separate function that fulfills the requirement of translation workflow. Although I do think that this is not an ideal solution to make it part of tokenizer and I would be happy to make API changes based on suggestions.

cpuhrsch commented on stl_text/ops/vocab/pytext_vocab.py:
```diff
@@ -1,28 +1,59 @@
+        >>> from stl_text.ops.vocab import VocabTransform
+        >>> vocab = VocabTransform("vocab.txt")
+        >>> jit_vocab = torch.jit.script(vocab)
+        >>> print(jit_vocab.code)
```
What is the reasoning behind printing the code here as well?

parmeet replied:
> What is the reasoning behind printing the code here as well?

Thanks for pointing this out. It was just part of Bootcamp task, but we really don't need to keep it. I will remove it from the final version.

cpuhrsch commented on examples/translation/train.py:
```diff
@@ -44,13 +29,13 @@ def main(fast_dev_run=True):
+    source_text_transform.build_vocab(map(lambda x:x['text'],wmt14["cs"][train]))
+    target_text_transform = WhitespaceTokenizer()
+    target_text_transform.build_vocab(map(lambda x:x['text'],wmt14["en"][train]))
+    source_vocab = source_text_transform.vocab.get_stoi()
```
Why use the stoi dictionary instead of the full class as it was before?

parmeet replied:
> Why use the stoi dictionary instead of the full class as it was before?

This change is compatible with previous version. get_stoi() on Vocab returns python dict which was also the case earlier https://github.com/hudeven/text/blob/9fc5b4962282d7e1dc8d5fa5eb3258f253981060/examples/translation/train.py#L27

parmeet committed:
Revert "seperating 'trainable' into seperate function in WhitespaceTokenizer"

This reverts commit db95a2d58fd9d9a504b3e442d399952542726fb4.

parmeet commented:
@cpuhrsch @zhangguanheng66 @hudeven 
I reverted the last change that tried to separate vocab building and numberization in WhitespaceTokenizer (it was anyway bit misleading what I did :P)

**Points of Consideration:**

1. WhitespaceTokenizer is used in three workflows doc_classification, contrastive_pretraining and translation. As of now, it has nice functionality to build vocab on the fly that can be done during data preparation/preprocessing stage (although not necessarily the right design as Tokenizer is holding more responsibility than it should ideally be)
 
2. Changing the 'trainable' part from numberization have side effects in all the above three workflows. If we were to separate vocab building from numberization (as originally discussed here https://github.com/hudeven/text/pull/10#discussion_r563836230), we need to make appropriate changes in all the above three workflows. 

3. Further note that the reason we need to make this separation or change 'on-the-fly' training behavior is to eventually make use of VocabTransform (https://github.com/hudeven/text/blob/9fc5b4962282d7e1dc8d5fa5eb3258f253981060/stl_text/ops/vocab/pytext_vocab.py#L20) inside WhitespaceTokenizer which isn't feasible currently because VocabTransform API doesn't support 'training' and require either vocab path or vocab list during initialization.
 
4. This leads me to conclusion that this change is slightly beyond scope of this PR where we intend to incorporate torchtext building blocks in WhitespaceTokenizer and doc_classification. Well, actually point 3 was also suppose to be part of this PR, but due to reason mentioned (incompatible APIs) it couldn't be done till we decide on new API design.

So if we all agree, we should perhaps work on this separation in separate PR :)

parmeet committed:
fixing some linter issues

parmeet committed:
fixing linter issues

parmeet commented:
closing it up at the moment due to uncertainty in proposed changes.
